src/FluentValidation/DefaultValidatorExtensions.cs

```diff
@@ -1047,6 +1046,19 @@ public static class DefaultValidatorExtensions {
 			return ruleBuilder.SetValidator(new ScalePrecisionValidator(scale, precision) { IgnoreTrailingZeros = ignoreTrailingZeros });
 		}
 
+		/// <summary>
+		/// Defines a count validator on the current rule builder, but only for collection properties.
+		/// Validation will fail if the number of elements is outside of the specified range. The range is inclusive.
+		/// </summary>
+		/// <typeparam name="T">Type of object being validated</typeparam>
+		/// <param name="ruleBuilder">The rule builder on which the validator should be defined</param>
+		/// <param name="min"></param>
+		/// <param name="max"></param>
+		/// <returns></returns>
+		public static IRuleBuilderOptions<T, ICollection> Count<T>(this IRuleBuilder<T, ICollection> ruleBuilder, int min, int max) {

```

Comment:
By forcing the generic return type to be `ICollection`, this means that any subsequent validators in the chain to act on ICollection rather than the original property type. If you change this to `TProperty` and add a generic constraint (`where TProperty : ICollection`), then it will preserve the original property type for subsequent validators in the chain:

```csharp
public static IRuleBuilderOptions<T, TProperty> Count<T, TProperty>(this IRuleBuilder<T, TProperty> ruleBuilder, int min, int max) where TProperty : ICollection 
```

Reply:
Good addition, I wasn't aware of this.